tests(rpc): Add fixed test vectors for `z_getsubtreesbyindex` from zcashd to zebra

oxarbitrage opened the pull request:
Depends-On: #7555

## Motivation

We want to test the `z_getsubtreesbyindex` against fixed test vectors obtained from zcashd and see if they match zebra.

Close https://github.com/ZcashFoundation/zebra/issues/7446

## Solution

The idea is to add json outputs from `zcashd` into the `zebra-test` crate and then create a test that could compare this fixed outputs with answers from zebrad made in real time.

In order to test this in high heights we need a pretty much synchronized cache state for zebra.

Locally, i run this test as:

```
export ZEBRA_CACHED_STATE_DIR="/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zebra"
export RUST_LOG="info"
cargo test rpc_z_getsubtreesbyindex_zcashd_test_vectors -- --nocapture --ignored
```

## Review

This is not ready yet because i don't have the right outputs from zcashd. In order to use the `getsubtreesbyindex` i had to reindex and that actually ended up downloading all the blocks again so i have a couple of days at least to be in sync again with zcashd.

But i pushed so other people can take a look to the overall structure and let me know if this is a good or bad idea. Thanks!

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

<!--
Is there anything missing from the solution?
-->

oxarbitrage committed:
add fixed test vectors for `z_getsubtreesbyindex`

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+    let network = Network::Mainnet;
+
+    let (mut zebrad, zebra_rpc_address) = if let Some(zebrad_and_address) =
+        spawn_zebrad_for_rpc(network, "fully_synced_rpc_test", test_type, true)?
```
Nitpick:

```suggestion
        spawn_zebrad_for_rpc(network, "rpc_z_getsubtreesbyindex_zcashd_test_vectors", test_type, true)?
```

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+    };
+
+    // Wait for zebrad to load the full cached blockchain, locally this can take up to 3 minutes.
+    std::thread::sleep(Duration::from_secs(180));
```
Waiting for a fixed amount of time is unreliable, because if the network or CI machine is slow, the test will fail. Instead, wait until Zebra logs that it is ready.

The test can still fail if it is upgrading an older cached state, so we might need to wait for that log as well. (Or maybe it's ok if all the cached states have been upgraded already. We can wait and see if it passes CI.)

```suggestion
    zebrad.expect_stdout_line_matches(SYNC_FINISHED_REGEX)?;
```

oxarbitrage replied:
I think this is a good suggestion and i appreciate it. I will apply, thanks.

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+    let zcashd_test_vectors = vec![
+        (
+            "test_mainnet_sapling_0_1.json".to_string(),
+            "[\"sapling\", 0, 1]".to_string(),
```
Optional: raw string literals will be easier to read here:
https://doc.rust-lang.org/rust-by-example/std/str.html#literals-and-escapes

This change can also be applied to Orchard.

```suggestion
            r#"["sapling", 0, 1]"#.to_string(),
```

teor2345 replied:
Edit: now that we've resolved my other suggestion, this one might be useful.

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+        // Parse results from RPC call
+        let body = res.bytes().await;
+        let parsed: Value = serde_json::from_slice(&body.expect("Response is valid json"))?;
+        let response_results = parsed["result"].to_string().replace(['\n', ' '], "");
```
Most of this loop seems like a manual re-implementation of `cargo insta`. It can automatically create and check JSON snapshot files in a single line of code. For example:
https://github.com/ZcashFoundation/zebra/blob/a876ddef67f0ff876015cda98d55094d6963f932/zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs#L581

Would it be easier to use `cargo insta` snapshots, then do a manual compare with `zcashd` using `zcash-rpc-diff`?
We'd only need to do the manual compare once, because the snapshots are checked into CI.

oxarbitrage replied:
This will not work. `zcash-rpc-diff` expects two running nodes. We don't have zcashd nodes in our CI and i don't think we will want them either.

oxarbitrage replied:
Also, if we want to do this we need to create snapshots for zcashd and store them. Then use `assert_json_snapshot` to compare them with what we get from zebrad. This means we need to write code for grabbing snapshots from zcashd which we currently don't have. I am not saying this is very hard but i thought about this option before and the code presented ended up being easier.

teor2345 replied:
> This will not work. `zcash-rpc-diff` expects two running nodes. We don't have zcashd nodes in our CI and i don't think we will want them either.

I'm sorry that I explained this badly, but that's not what I meant at all.

I meant that we do two different things with the same Zebra code:
1. Use `cargo insta` to take snapshots which we automatically check in the tests.
2. Do one manual check that the RPC responses we are snapshotting match `zcashd`, then never check again.

The manual check with `zcash-rpc-diff` does not need to happen in CI, because the snapshots from the same Zebra version are already checked in CI.

But it's your decision, so I will resolve everything related to this suggestion.

oxarbitrage replied:
That makes more sense, thank you for the clarification.

If you prefer that i can try it tomorrow. For me is the same but more work changing it out.

teor2345 replied:
If you can get the test working the way it is, then there's no need to change it.

I don't expect the RPC format to change very often. (It might change once to add start heights, but that's it.)

oxarbitrage replied:
What is the issue you see the way it is that could be better with the snapshot option?

Edit: question is referring to making the tests work

teor2345 replied:
Thanks, that's a good question! When I do a review, I try to understand what every line of code does, and why it is there.

I don't think there's anything had about doing it this way, or wrong in this code. So I think you're right, and it's not worth changing. I was mainly thinking about how much effort it would be to review. (I still think it's ok, it would probably take me half an hour to review, rather than 5 minutes.)

I'm mainly sharing these details because you asked how I review things, but you don't need to read them if you're busy:

<details>

<summary>how I would do this review</summary>

If we were using `insta` here, there would be one line of code to review here, that I have seen before. So it is easy to know what it does. (I know how `insta` snapshots work in tests, and they are very reliable.) So that would pretty much be the end of my review.

But in this test I have to check 10 lines of new code, and make sure they test what is in the ticket. So I might ask myself questions like:
* why do we only want the `result` from the JSON, and is that the right thing to select? (Probably ok.)
* why do we need to remove whitespace, and are spaces and newlines the only whitespace we care about? Are we removing the same whitespace for the test vectors and the live responses? (Seems ok, but needs to be looked at by the reviewer.)
* is the test file path correct? Will it work in CI and on people's local machines? (Maybe, I'll have to run it to check.)
* will the test file path work on Windows / macOS? (Maybe  / Yes, but it doesn't really matter because if it fails it's ok to skip this test.)

These issues apply both to this code and if we used `insta`, but there's more lines to check in this code:
* are there ways that this test can fail or be unreliable?
* are there ways that this test can fail to run and still look like it succeeded? (Yes, the `#[ignore]` could accidentally skip the test if we don't use the right command in the test `entrypoint.sh`. Otherwise it looks like the checks always run when there's a cached state.)
* do the snapshots match what `zcashd` produces? (Needs to be checked by a reviewer manually.)

</details>

oxarbitrage replied:
I think what you never consider in your reviews is how much effort it will take the developer to make all the changes you want.

The way you have to do it (very rational, always with the right answer for everything) discourages me personally to make pull requests that are greater than one line of code.

oxarbitrage replied:
How told you this was the PR to make review lessons ? This is a daily routine PR that was assigned to me by my project manager (i have pressure to finalize it) among with other two that i can't merge for CI issues. You need to add context to your reviews as well.

oxarbitrage replied:
Why don't you spend your time helping me with the other 2 PR that i requested your help instead of discussing in this one which is probably the less important ? Also the one that do not have any major issue, which is ok as it is (said by you) but ....

Help me with the ones that nobody else can and i get this one sorted out with any other member of the team.

teor2345 replied:
I thought we were having a conversation about alternative test designs because you were interested. But I can see that you want it to stop. So I will stop after this comment.

> Why don't you spend your time helping me with the other 2 PR that i requested your help

It is the middle of my working day. I have time to look the structure of this PR like you asked, and do a detailed review of the other two PRs. If you don't have work time now, that's ok. I can do it after I have lunch, and you will see it tomorrow.

Next time you want to switch my priorities around, please DM me and Pili.

teor2345 reviewed (dismissed):
This looks like it will work and cover the test cases that we want it to. But I wonder if there's an easier way using `cargo insta`?

If we use snapshots here instead, we can create and maintain them using the command:
`ZEBRA_CACHED_STATE_DIR="/path/to/zebra/state" cargo insta test --review --delete-unreferenced-snapshots`

Then we just need to check these test cases once with `zcash-rpc-diff`. I can do that by adding them to ticket #7450.

teor2345 commented:
Just to be clear: I think the way this PR does the test manually with test vectors should also work. But it might take more effort to get working, or to maintain if the RPC changes.

oxarbitrage commented:
> Just to be clear: I think the way this PR does the test manually with test vectors should also work. But it might take more effort to get working, or to maintain if the RPC changes.

I think this is the wrong way to see it. The effort to make it work is already done. I can accept that it could be harder to maintain if the rpc changes.

There is no easier way than the one i already did. There are better/different ways but none of them is easier for me.

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+    std::thread::sleep(Duration::from_secs(180));
+
+    // Create an http client
+    let client = RpcRequestClient::new(zebra_rpc_address.expect(""));
```
```suggestion
    let client = RpcRequestClient::new(zebra_rpc_address.expect("already checked that address is valid"));
```

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
+/// Require scyncronized zebrad cache state to query high heights.
+#[tokio::test]
+#[ignore]
+async fn rpc_z_getsubtreesbyindex_zcashd_test_vectors() -> Result<()> {
```
This is a note about running this test from the `entrypoint.sh` file, but I can't make suggestions there using GitHub:

We need to run this test with a cached Zebra state, but we're also trying to limit the number of CI jobs we have.

So I suggest adding it as another test in the same job as the existing Zebra RPC test:
https://github.com/ZcashFoundation/zebra/blob/5b0bc700981c39989f46a099d328e00b2166ebac/docker/entrypoint.sh#L88

You could copy that one line and change the test name, then make sure two test are being run in that CI job.

oxarbitrage committed:
change to snapshots

oxarbitrage committed:
add test to docker

oxarbitrage committed:
remove assert lines from snapshots

oxarbitrage committed:
add more tests

oxarbitrage committed:
Merge remote-tracking branch 'origin/main' into issue7446

oxarbitrage commented:
I added the tests mentioned [here](https://github.com/ZcashFoundation/zebra/issues/7446#issuecomment-1718645368) to this PR. 

All the responses for the test vectors added by now are the same as zcashd:

```
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex sapling 0 1
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex sapling 0 1

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,002s
user	0m0,002s
sys	0m0,000s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,006s
user	0m0,005s
sys	0m0,000s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.R63LEBv9BJ.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "sapling",
  "start_index": 0,
  "subtrees": [
    {
      "root": "754bb593ea42d231a7ddf367640f09bbf59dc00f2c1d2003cc340e0c016b5b13",
      "end_height": 558822
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex sapling 0 11
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex sapling 0 11

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,002s
user	0m0,001s
sys	0m0,000s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,002s
user	0m0,000s
sys	0m0,001s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.g132P9v08V.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "sapling",
  "start_index": 0,
  "subtrees": [
    {
      "root": "754bb593ea42d231a7ddf367640f09bbf59dc00f2c1d2003cc340e0c016b5b13",
      "end_height": 558822
    },
    {
      "root": "03654c3eacbb9b93e122cf6d77b606eae29610f4f38a477985368197fd68e02d",
      "end_height": 670209
    },
    {
      "root": "e2bf698f5ac10b44da560d11a5e1d5c191a82a968a2be0a6948aa8748b545160",
      "end_height": 780364
    },
    {
      "root": "e71be1fd3963a2d700ed374d01fdd4a70d8dda8189a8f6602e5fe66c2c66a11d",
      "end_height": 839994
    },
...
      "end_height": 1056197
    },
    {
      "root": "163838d11a0525898f410397ae6ec627fa18b3184a021e088806947b3c5ca718",
      "end_height": 1109360
    },
    {
      "root": "089a1f9d50a037cc66aba4400b1703bcbb66f5f2993fd0dd3bb726e359409167",
      "end_height": 1169471
    },
    {
      "root": "0fc8a6ad74e65764d9bc2fcd13866b994731dfda821e6618a1a308bb1eccf51a",
      "end_height": 1256932
    },
    {
      "root": "cdd92a1e884cf1914dae8345423203832ec7bbf9d95ae50b82e4327b39d6d912",
      "end_height": 1363036
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex sapling 17 1
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex sapling 17 1

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,004s
user	0m0,004s
sys	0m0,000s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,002s
user	0m0,002s
sys	0m0,000s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.QvmdYXCO7r.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "sapling",
  "start_index": 17,
  "subtrees": [
    {
      "root": "89ac51fcf846eacb2e7bca32d9bf3793792d2d952c1fec11e52b56d7674c1c36",
      "end_height": 1703171
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex sapling 1090 6
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex sapling 1090 6

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,005s
user	0m0,000s
sys	0m0,004s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,030s
user	0m0,003s
sys	0m0,000s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.R7U1U2vWZZ.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "sapling",
  "start_index": 1090,
  "subtrees": [
    {
      "root": "507ebccc63e2ca45bd254975ca2b75f2a221a0167d514077472dc7ef1e04a638",
      "end_height": 1857525
    },
    {
      "root": "0240b333e243a38583b692b9ce5b2fbdfde41cfdd3f3067b5c7d94fda5387666",
      "end_height": 1860908
    },
    {
      "root": "bf299f8446bea806854e9b363ed890f83423cececd53293c81df92ff112aa43f",
      "end_height": 1866377
    },
    {
      "root": "4a4778e6242b3e708f040b4a373c8cdedd14c5c415ab20f32e46430507b1a346",
      "end_height": 1889548
    },
    {
      "root": "c1cffa22d637cb194102a1b02a57d2b47c67f54094f6773a38b2e53d3becb45f",
      "end_height": 1997495
    },
    {
      "root": "0ee7328354ad588ee9581b25a3e2b94dbb03db647b79538c8805d793c61b822d",
      "end_height": 2056616
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex orchard 0 1
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex orchard 0 1

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,002s
user	0m0,002s
sys	0m0,000s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,002s
user	0m0,001s
sys	0m0,000s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.Yijqpe8TRh.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "orchard",
  "start_index": 0,
  "subtrees": [
    {
      "root": "d4e323b3ae0cabfb6be4087fec8c66d9a9bbfc354bf1d9588b6620448182063b",
      "end_height": 1707429
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex orchard 338 1
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228239

Request:
z_getsubtreesbyindex orchard 338 1

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,002s
user	0m0,000s
sys	0m0,001s

Querying zcashd v5.6.1 main chain at height >=2228239...

real	0m0,002s
user	0m0,001s
sys	0m0,000s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.aZVC2z7zpl.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "orchard",
  "start_index": 338,
  "subtrees": [
    {
      "root": "52c8d1ea6ef49c6e0d6bb6fef4520e1e3851895a04b52bfac1b1cc0fbbb3f709",
      "end_height": 1888929
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ ./zcash-rpc-diff 6666 z_getsubtreesbyindex orchard 585 1
Checking first node release info...
Checking second node release info...
Connected to zebrad v1.2.0+24.g199c2bb.modified (port 6666) and zcashd v5.6.1 (./zcash-cli -rpccookiefile=/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash/.cookie zcash.conf port).

Checking zebrad v1.2.0+24.g199c2bb.modified network and tip height...
Checking zcashd v5.6.1 network and tip height...

WARNING: comparing RPC responses from different heights:
zebrad v1.2.0+24.g199c2bb.modified is at: 2228238
zcashd v5.6.1 is at: 2228240

Request:
z_getsubtreesbyindex orchard 585 1

Querying zebrad v1.2.0+24.g199c2bb.modified main chain at height >=2228238...

real	0m0,002s
user	0m0,001s
sys	0m0,000s

Querying zcashd v5.6.1 main chain at height >=2228240...

real	0m0,002s
user	0m0,000s
sys	0m0,001s


Response diff between zebrad v1.2.0+24.g199c2bb.modified and zcashd v5.6.1:
RPC responses were identical

/tmp/tmp.9jfRh8iB2V.rpc-diff/zebrad-main-2228238-z_getsubtreesbyindex.json, limited to 40 lines:
{
  "pool": "orchard",
  "start_index": 585,
  "subtrees": [
    {
      "root": "43c86869520ada7a2cd0deeecf3650f337f90bd4328b16144b0a278ad9fdaa08",
      "end_height": 2000126
    }
  ]
}
oxarbitrage@oxarbitrage:/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zcash$ 

```

oxarbitrage committed:
change test description

oxarbitrage committed:
change test name

teor2345 commented:
The fully-synced-rpc failure looks like bug #7564:
> tee: 'standard output': Broken pipe

https://github.com/ZcashFoundation/zebra/actions/runs/6207948275/job/16854306976?pr=7515#step:9:280

teor2345 commented:
The `run-lwd-grpc-wallet` failure seems to be caused by an invalid cached state from a previous CI run:
> 2023-09-16T15:57:12.977740Z  INFO zebra_state::service::finalized_state::disk_format::upgrade: trying to open current database format running_version=Version { major: 25, minor: 2, patch: 0 }
> ...
> 2023-09-16T15:57:19.426808Z ERROR zebra_state::service::finalized_state::disk_format::upgrade::add_subtrees: missing subtree index=564

https://github.com/ZcashFoundation/zebra/actions/runs/6207948275/job/16854295441?pr=7515#step:9:185

That cached state is:
> Description: Created from commit c187e2a with height 2227996 and database format v25-2-0 changing to v25-2-1
> Selected Disk: lwd-cache-7566-merge-c187e2a-v25-mainnet-tip-u-152354

Which means the bug is in PR #7566. I'll fix the bug and delete those cached states.

teor2345 commented on docker/entrypoint.sh:
```diff
@@ -86,6 +86,7 @@ case "$1" in
             # Starting at a cached Zebra tip, test a JSON-RPC call to Zebra.
             ls -lh "$ZEBRA_CACHED_STATE_DIR"/*/* || (echo "No $ZEBRA_CACHED_STATE_DIR/*/*"; ls -lhR  "$ZEBRA_CACHED_STATE_DIR" | head -50 || echo "No $ZEBRA_CACHED_STATE_DIR directory")
             cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_test
+            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored rpc_z_getsubtreesbyindex_sync_snapshots
```
I think the test runner is stopping after the first test, due to this workflow code:
https://github.com/ZcashFoundation/zebra/blob/a7501c16b1c9a3531f3cf833b10d1fcdb98f6393/.github/workflows/deploy-gcp-tests.yml#L597-L599

This might also be the cause of bug #7564.

So it might be easier to run both tests at the same time. That will need a change to the test name, I'll make another suggestion below.

```suggestion
            # Run both the fully synced RPC test and the subtree snapshot test
            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_
```

oxarbitrage replied:
Thanks. https://github.com/ZcashFoundation/zebra/pull/7515/commits/2ebb28032288e41ea80176b1ddcee59208cfcb28

teor2345 commented on zebrad/Cargo.toml:
```diff
@@ -235,6 +235,7 @@ hex-literal = "0.4.1"
 jsonrpc-core = "18.0.0"
 once_cell = "1.18.0"
 regex = "1.9.5"
+insta = { version = "1.31.0", features = ["json"] }
```
Optional Nitpick: Is this dependency used anywhere?

oxarbitrage replied:
https://github.com/ZcashFoundation/zebra/blob/33987ed8cb040bf56d3f93cf545873873d62b758/zebrad/tests/acceptance.rs#L2676

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2605,86 @@ async fn state_format_test(
+/// Snapshot the `z_getsubtreesbyindex` method in a synchronized chain.
+#[tokio::test]
+#[ignore]
+async fn rpc_z_getsubtreesbyindex_sync_snapshots() -> Result<()> {
```
This name change is part of running both tests at the same time.

```suggestion
/// Snapshot the `z_getsubtreesbyindex` method in a synchronized chain.
///
/// This test name must have the same prefix as the `fully_synced_rpc_test`, so they can be run in the same test job.
#[tokio::test]
#[ignore]
async fn fully_synced_rpc_z_getsubtreesbyindex_snapshot_test() -> Result<()> {
```

oxarbitrage replied:
Done at https://github.com/ZcashFoundation/zebra/pull/7515/commits/2ebb28032288e41ea80176b1ddcee59208cfcb28

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2605,86 @@ async fn state_format_test(
+        // Skip the test, we don't have the required cached state
+        return Ok(());
+    };
+
```
Since this test requires the subtree upgrade to be complete, we'll need to wait for it.

These functions are added in PR #7566, so you'll need to wait for it to merge, or copy them across to this PR.

This code gets the log message that says if an upgrade is running, it needs to be run after `zebrad` is launched:
https://github.com/ZcashFoundation/zebra/blob/be6a33ee6598f7cd86cbd1da7d961a606c88d07b/zebrad/tests/common/lightwalletd/wallet_grpc_test.rs#L102-L103

teor2345 commented on zebrad/tests/acceptance.rs:
```diff
@@ -2606,3 +2605,86 @@ async fn state_format_test(
+
+    // Wait for zebrad to load the full cached blockchain.
+    zebrad.expect_stdout_line_matches(SYNC_FINISHED_REGEX)?;
+
```
This code waits for the subtree upgrade to finish, it needs to be run before we test subtrees:
https://github.com/ZcashFoundation/zebra/blob/be6a33ee6598f7cd86cbd1da7d961a606c88d07b/zebrad/tests/common/lightwalletd/wallet_grpc_test.rs#L387-L397

teor2345 reviewed (dismissed):
Looks good!

There's a few tricky things about how the test is run, and when we run it. I think I've fixed them in PR #7566, but we'll need to do something similar here.

oxarbitrage committed:
run both tests together

oxarbitrage committed:
Merge remote-tracking branch 'origin/main' into issue7446

oxarbitrage committed:
wait for state version update in test

oxarbitrage commented:
Locally this test hanged with the latest changes but i want to check in the CI.

teor2345 commented on docker/entrypoint.sh:
```diff
@@ -85,7 +85,8 @@ case "$1" in
             ls -lh "$ZEBRA_CACHED_STATE_DIR"/*/* || (echo "No $ZEBRA_CACHED_STATE_DIR/*/*"; ls -lhR  "$ZEBRA_CACHED_STATE_DIR" | head -50 || echo "No $ZEBRA_CACHED_STATE_DIR directory")
-            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_test
+            # Run both the fully synced RPC test and the subtree snapshot test
+            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_
```
I'm really sorry about this, but my suggestion was incomplete.

What's happening now is that both tests are running in parallel on the same cached state directory, so one of them panics with a lock file error:
> Message:  Opening database "/var/cache/zebrad-cache/state/v25/mainnet" failed: Error { message: "IO error: While lock file: /var/cache/zebrad-cache/state/v25/mainnet/LOCK: Resource temporarily unavailable" }. Hint: Check if another zebrad process is running. Try changing the state cache_dir in the Zebra config.

https://github.com/ZcashFoundation/zebra/actions/runs/6241663187/job/16945479607?pr=7515#step:9:232

So far I've suggested two different ways to add an extra cached state test, because I thought they would be faster for you. But we've never done it this way before, and I didn't check myself. So that's my mistake.

I will try one last thing in this PR and see if it works. Otherwise I'll open a ticket to document the standard way to add a new job.

teor2345 replied:
If CI still isn't passing after my fix, here are the standard "new test" steps from ticket #7589.

But we're seeing CI failures for other reasons right now, so maybe it's worth me double-checking the CI output before you do all this work.

<details>

<summary>new job steps: don't do this unless CI fails and we can't fix it</summary>

To add a new cached state test job to CI:
- [x] Add a new Rust test and mark it as `#[ignored]`
- [ ] Run that Rust test from a new `elif` block in [entrypoint.sh](https://github.com/ZcashFoundation/zebra/blob/main/docker/entrypoint.sh)
- [ ] Add a new job to `[continous-integration-docker.yml](https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/continous-integration-docker.yml)
- [ ] Make the `failure-issue` job depend on the new job
- [ ] Add a new patch job to [continous-integration-docker.patch.yml](https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/continous-integration-docker.patch.yml)
- [ ] Check the new test is being run by looking at the CI logs
- [ ] Admin: after the PR merges, add a new branch protection rule for the new job

</details>

teor2345 commented on docker/entrypoint.sh:
```diff
@@ -85,7 +85,8 @@ case "$1" in
             ls -lh "$ZEBRA_CACHED_STATE_DIR"/*/* || (echo "No $ZEBRA_CACHED_STATE_DIR/*/*"; ls -lhR  "$ZEBRA_CACHED_STATE_DIR" | head -50 || echo "No $ZEBRA_CACHED_STATE_DIR directory")
-            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_test
+            # Run both the fully synced RPC test and the subtree snapshot test
+            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored fully_synced_rpc_
```
```suggestion
            # Run both the fully synced RPC test and the subtree snapshot test, one test at a time.
            # Since these tests use the same cached state, a state problem in the first test can fail the second test.
            cargo test --locked --release --features "$ENTRYPOINT_FEATURES" --package zebrad --test acceptance -- --nocapture --include-ignored --test-threads 1 fully_synced_rpc_
```

teor2345 committed:
Run one test at a time to avoid state locking issues

teor2345 reviewed (approved):
Looks like that worked!
